iOS_SDK/OneSignalSDK/Source/OSMessagingController.m

```diff
@@ -295,33 +298,48 @@ - (void)messageViewControllerWasDismissed {
             // Remove dismissed IAM from messageDisplayQueue
             [self.messageDisplayQueue removeObjectAtIndex:0];
         }
-        
         // Reset the IAM viewController to prepare for next IAM if one exists
         self.viewController = nil;
         // No IAMs are showing currently
         self.isInAppMessageShowing = false;
         // Reset time since last IAM
         [self.triggerController timeSinceLastMessage:[NSDate new]];
         
-        if (self.messageDisplayQueue.count > 0) {
-            // Show next IAM in queue
-            [self displayMessage:self.messageDisplayQueue.firstObject];
-            return;
+        if (_promptAction) {
+            [_promptAction handlePrompt:^(BOOL accepted) {
+                self.promptAction = nil;
+                [self evaluateMessageDisplayQueue];
+            }];
         } else {
-            // Hide the window and call makeKeyWindow to ensure the IAM will not be shown
-            self.window.hidden = true;
-            [UIApplication.sharedApplication.delegate.window makeKeyWindow];
-            
-            // Evaulate any IAMs (could be new IAM or added trigger conditions)
-            [self evaluateMessages];
+            [self evaluateMessageDisplayQueue];
         }
     }
 }
 
+- (void)evaluateMessageDisplayQueue {
+    if (self.messageDisplayQueue.count > 0) {
+        // Show next IAM in queue
+        [self displayMessage:self.messageDisplayQueue.firstObject];
+        return;
+    } else {
+        // Hide the window and call makeKeyWindow to ensure the IAM will not be shown
+        self.window.hidden = true;
+        [UIApplication.sharedApplication.delegate.window makeKeyWindow];
+       
+        // Evaulate any IAMs (could be new IAM or added trigger conditions)
+        [self evaluateMessages];
+    }
+}
+
 - (void)messageViewDidSelectAction:(OSInAppMessage *)message withAction:(OSInAppMessageAction *)action {
     // Assign firstClick BOOL based on message being clicked previously or not
     action.firstClick = [message takeActionAsUnique];
     
+    if (action.promptAction) {
+        self.promptAction = action.promptAction;
+        [self.viewController dismissCurrentInAppMessage];
```

Comment:
Lets give the IAM a dismiss action in the JS, HTML, CSS code and we can remove this dismissal